core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxBuilder.scala

```diff
@@ -0,0 +1,263 @@
+package org.bitcoins.core.wallet.builder
+
+import org.bitcoins.core.number.{Int32, UInt32}
+import org.bitcoins.core.protocol.script.{
+  CLTVScriptPubKey,
+  CSVScriptPubKey,
+  EmptyScriptSignature
+}
+import org.bitcoins.core.protocol.transaction.{
+  BaseTransaction,
+  EmptyWitness,
+  NonWitnessTransaction,
+  Transaction,
+  TransactionConstants,
+  TransactionInput,
+  TransactionOutput,
+  WitnessTransaction
+}
+import org.bitcoins.core.wallet.utxo.{
+  ConditionalInputInfo,
+  EmptyInputInfo,
+  InputInfo,
+  InputSigningInfo,
+  LockTimeInputInfo,
+  MultiSignatureInputInfo,
+  P2PKHInputInfo,
+  P2PKInputInfo,
+  P2PKWithTimeoutInputInfo,
+  P2SHInputInfo,
+  P2WPKHV0InputInfo,
+  P2WSHV0InputInfo,
+  UnassignedSegwitNativeInputInfo
+}
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+class RawTxBuilder {
+  private var version: Int32 = TransactionConstants.validLockVersion
+
+  private val inputsBuilder: mutable.Builder[
+    TransactionInput,
+    Vector[TransactionInput]] = Vector.newBuilder
+
+  private val outputsBuilder: mutable.Builder[
+    TransactionOutput,
+    Vector[TransactionOutput]] = Vector.newBuilder
+
+  private var lockTime: UInt32 = TransactionConstants.lockTime
+
+  private var finalizerOpt: Option[RawTxFinalizer] = None
+
+  def hasFinalizer: Boolean = finalizerOpt.isDefined
+
+  def setFinalizer(finalizer: RawTxFinalizer): this.type = {
+    finalizerOpt = Some(finalizer)
+    this
+  }
+
+  def result()(implicit ec: ExecutionContext): Future[Transaction] = {
+    finalizerOpt match {
+      case None =>
+        Future.failed(
+          new RuntimeException(
+            "Cannot call result with no RawTxFinalizer specified"))
+      case Some(finalizer) => result(finalizer)
+    }
+  }
+
+  def result(finalizer: RawTxFinalizer)(
+      implicit ec: ExecutionContext): Future[Transaction] = {
+    finalizer.buildTx(version,
+                      inputsBuilder.result(),
+                      outputsBuilder.result(),
+                      lockTime)
+  }
+
+  def clear(): Unit = {
+    version = TransactionConstants.validLockVersion
+    inputsBuilder.clear()
+    outputsBuilder.clear()
+    lockTime = TransactionConstants.lockTime
+  }
+
+  def addInput(input: TransactionInput): this.type = {
```

Comment:
so are there any requirements onthese `TransactionInput`? For instance, do they need to have `EmptyScriptSignature` be the scriptSig? Can the ScriptSignature have the redeemScript in the case case of p2sh?

Reply:
Depending on the answer above, we don't seem to have the ability to add any `ScriptWitness` to the transaction. Should we have a `+=(tuple: (TransactionInput,ScriptWitness)) method?